### ISSUE TITLE
Android system removed permissions and c:geo does not request them

### ISSUE THREAD

**Issue by andrixnet** [cgeo/cgeo#14069]
### Describe your problem!

I've been away from geocaching for several months due to health reasons and I haven't used c:geo in a while. 
My phone from time to time notifies me that "it removed permissions from unused apps" (gracefully saying nothing about which apps were those; <rant> thanks benevolent Google </rant> ). 

Today I started c:geo and it kept not showing my offline maps. At first I thought it may be a problem with the maps, but when trying online tiles, it showed me in Paris (obviously wrong) and it didn't update even after confirming GPS fix with another app. I even downloaded another version of my country's map using c:geo own support for it. It did download the file (which is odd considering below).

On a hunch, I went to App info -> permissions. 
There was "location" and "storage", both off. 
I turned them both on and magically c:geo is working again. 

### How to reproduce?

Have Android system remove permissions from c:geo due to lack of use. 
or you could just turn off the permissions from Settings -> Apps -> permissions. 

### Actual result after these steps?

c:geo starts, runs apparently ok, but will place you in a fixed position in Paris, won't update location from GPS.

And most importantly **it will say nothing**.

### Expected result after these steps?

I know from experience from other apps that had their permissions removed (by the system, due to lack of use) that when they are first started after such an event, they will ask for permission just like the first run after installation. 

c:geo has permissions off and says nothing about it.
I expected c:geo to ask me to grant it the permission it actually needs to properly run, like other apps do in this case.

### Reproducible

Yes

### c:geo Version

2023.02.25

### System information

```text
## System information

c:geo version: 2023.02.25

Device:- Device type: SM-G965F (star2ltexx, samsung)
- Available processors: 8
- Android version: 9
- Android build: PPR1.180610.011.G965FXXS7CSK4
- Screen resolution: 1440x2792px (411x773dp)
- Pixel density: 3.8500001
- System font scale: 1.1 / used scale: 1.1
- Sailfish OS detected: false
- Google Play services: enabled - 23.07.14 (100400-512733570)
- Memory: Available:1.8 GB, Total:5.5 GB, Threshold: 216.0 MB, low:false

Sensor and location:- Low power mode: inactive
- Compass capabilities: yes
- Rotation vector sensor: present
- Orientation sensor: present
- Magnetometer & Accelerometer sensor: present
- Direction sensor used: rotation vector

Program settings:- Settings: v9, Count:194
- Set language: en_GB (system default)
- System date format: dd/MM/y
- Time zone: GMT+02:00
- Debug mode active: no
- Last backup: 4 Nov 2022, 16:42
- Routing mode: Bike
- Live map mode: false
- OSM multi-threading: true / threads: 4
- Map: Romania_v5.2021-12-30 (Offline)
- Id: cgeo.geocaching.maps.mapsforge.MapsforgeMapProvider$OfflineMapSource:primary:maps/mapsforge/romania_v5.2021-12-30.map
- Atts: Map data (c) OpenStreetMap contributors
- Theme: none

Filters:- Hide waypoints: -
- LIVE: Origin: geocaching.com, Status: 3 items ([:inconclusive=false:advanced=true]AND(stored_list;attributes:inverse=false:attributesources=3;origin:geocaching.com;status:found_no:exclude_disabled:exclude_archived))
- OFFLINE: None ([:inconclusive=false:advanced=false]AND(type;difficulty_terrain:d=-:d=-:t=-:t=-;status))

Stored Filters (#0):

Services:- Geocaching sites enabled:
geocaching.com: Logged in (Login OK) / BASIC
opencaching.ro: Logged in (Login OK)
- Geocaching.com date format: yyyy-MM-dd
- Geocaching.com website language: en-US
- Last login error on geocaching.com: communication error (2022-08-20 22-15)
- Last successful login on geocaching.com: 2023-03-08 21-29
- Routing: external / BRouter installed: true
- Installed c:geo plugins: none

Permissions-android.permission.ACCESS_COARSE_LOCATION:: DENIED
-android.permission.ACCESS_FINE_LOCATION:: DENIED
-android.permission.READ_CONTACTS:: DENIED
-android.permission.READ_EXTERNAL_STORAGE:: DENIED
-android.permission.WRITE_EXTERNAL_STORAGE:: DENIED

Paths- System internal c:geo dir: /data/user/0/cgeo.geocaching (179.9 GB free) v3 internal isDir(8 entries)
- Legacy User storage c:geo dir: /data/user/0/cgeo.geocaching (179.9 GB free) v3 internal isDir(8 entries)
- Geocache data: /storage/emulated/0/Android/data/cgeo.geocaching/files/GeocacheData (179.9 GB free) v3 external non-removable isDir(421 entries)
- Internal theme sync (is turned off): /data/user/0/cgeo.geocaching/MapThemeData (179.9 GB free) v3 internal isDir(0 entries)
- Public Folders: #12
- BASE: /cgeo (User-Defined)[/cgeo[DOCUMENT#0:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo, Av:true, files:>=6, dirs:>=15, totalFileSize:>=37.2 MB, free space: 179.9 GB, files on device: 7598080)
- OFFLINE_MAPS: /maps/mapsforge (User-Defined)[/maps/mapsforge[DOCUMENT#0:p-content://com.android.externalstorage.documents/tree/primary%3Amaps%2Fmapsforge::]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Amaps%2Fmapsforge/document/primary%3Amaps%2Fmapsforge, Av:true, files:5, dirs:2, totalFileSize:789.8 MB, free space: 179.9 GB, files on device: 7598080)
- OFFLINE_MAP_THEMES: /maps/mapsforge (User-Defined)[/maps/mapsforge[DOCUMENT#0:p-content://com.android.externalstorage.documents/tree/primary%3Amaps%2Fmapsforge::]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Amaps%2Fmapsforge/document/primary%3Amaps%2Fmapsforge, Av:true, files:5, dirs:2, totalFileSize:789.8 MB, free space: 179.9 GB, files on device: 7598080)
- OFFLINE_MAP_SHADING: /maps/mapsforge/_hgt (Default)[/maps/mapsforge/_hgt[PERSISTABLE_FOLDER(OFFLINE_MAPS)#1:p-content://com.android.externalstorage.documents/tree/primary%3Amaps%2Fmapsforge::/_hgt]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Amaps%2Fmapsforge/document/primary%3Amaps%2Fmapsforge%2F_hgt, Av:true, files:0, dirs:0, totalFileSize:0 B, free space: 179.9 GB, files on device: 7598080)
- LOGFILES: /cgeo/logfiles (Default)[/cgeo/logfiles[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/logfiles]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Flogfiles, Av:true, files:0, dirs:0, totalFileSize:0 B, free space: 179.9 GB, files on device: 7598080)
- GPX: /gpx (User-Defined)[/gpx[DOCUMENT#0:p-content://com.android.externalstorage.documents/tree/primary%3Agpx::]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Agpx/document/primary%3Agpx, Av:true, files:4, dirs:0, totalFileSize:6.4 MB, free space: 179.9 GB, files on device: 7598080)
- BACKUP: /cgeo/backup (Default)[/cgeo/backup[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/backup]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Fbackup, Av:true, files:>=8, dirs:>=8, totalFileSize:>=49.6 MB, free space: 179.9 GB, files on device: 7598080)
- FIELD_NOTES: /cgeo/field-notes (Default)[/cgeo/field-notes[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/field-notes]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Ffield-notes, Av:true, files:0, dirs:0, totalFileSize:0 B, free space: 179.9 GB, files on device: 7598080)
- SPOILER_IMAGES: /cgeo/GeocachePhotos (Default)[/cgeo/GeocachePhotos[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/GeocachePhotos]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2FGeocachePhotos, Av:true, files:0, dirs:0, totalFileSize:0 B, free space: 179.9 GB, files on device: 7598080)
- ROUTING_BASE: /cgeo/routing (Default)[/cgeo/routing[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/routing]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Frouting, Av:true, files:8, dirs:1, totalFileSize:48.9 MB, free space: 179.9 GB, files on device: 7598080)
- ROUTING_TILES: /cgeo/routing/segments4 (Default)[/cgeo/routing/segments4[PERSISTABLE_FOLDER(ROUTING_BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/routing/segments4]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Frouting%2Fsegments4, Av:true, files:1, dirs:0, totalFileSize:48.8 MB, free space: 179.9 GB, files on device: 7598080)
- TEST_FOLDER: [Legacy]/data/user/0/cgeo.geocaching/files/unittest (Default)[/data/user/0/cgeo.geocaching/files/unittest[FILE#1:p-file:///data/user/0/cgeo.geocaching/files::/unittest]] (Uri: file:///data/user/0/cgeo.geocaching/files/unittest, Av:true, files:0, dirs:0, totalFileSize:0 B, free space: 179.9 GB, files on device: -1)
- Map render theme path:
- PersistedDocumentUris: #1
- UNUSED_TRACK: null
- Persisted Uri Permissions: #3
- content://com.android.externalstorage.documents/tree/primary%3Amaps%2Fmapsforge (19 Apr 2021, 11:32):RW
- content://com.android.externalstorage.documents/tree/primary%3Agpx (19 Apr 2021, 11:32):RW
- content://com.android.externalstorage.documents/tree/primary%3Acgeo (19 Apr 2021, 11:32):RW
- Database: /storage/emulated/0/Android/data/cgeo.geocaching/files/databases/data (v99, Size:10.9 MB) on user storage

--- End of system information ---
```

### Additional Information

My c:geo was at an earlier version initially, when I first encountered the problem, had it update (via Google Play), tried again with the latest version, of course the same problem.

The system info shown here is with location and storage permissions off.

**Comment by eddiemuc** [comment]
c:geo should show a warning on home screen for removed GPS:

![image](https://user-images.githubusercontent.com/6909759/223840181-079bcf6a-4cd3-4ae5-bbfc-34ea2a3315f6.png)

One could argue if this warning should be more prominent of course. 

Wrt storage permission, it should no longer be necessary for c:geo usage since it uses SAF completely. We had some strange behaviour with this in the past though. Will c:geo work even with removed storage permission?

**Comment by fm-sys** [comment]
> Wrt storage permission, it should no longer be necessary for c:geo usage since it uses SAF completely. We had some strange behaviour with this in the past though. Will c:geo work even with removed storage permission?

A bit off-topic, but today I tried to import a GPX track but nothing happened (even no error message!) 

I created a log file and found following lines:
```
03-10 13:05:51.921 18010 18433 W cgeo    : [RxCachedThreadScheduler-8] Could not make directories /storage/emulated/0/cgeo
03-10 13:05:51.921 18010 18433 W cgeo    : [RxCachedThreadScheduler-8] External public cgeo directory '/storage/emulated/0/cgeo' not available 
03-10 13:05:56.258 18010 18433 I cgeo    : [RxCachedThreadScheduler-8] [LogCat]Issuing command: [logcat, -d, AndroidRuntime:E, cgeo:D, cgeo.geocachin:I, *:S, -f, /data/user/0/cgeo.geocaching/cache/cgeo_tempfile_5131573082448292337.tmp]
```

What happened?:
My phone is running android 10 (API 29) and cgeo had lost its storage permission (a similar case was described in #13988). Since PR #14006 storage permission will be re-requested but not for API 29.

Three things that came to my mind:
- We should show a proper error message
- Why do we need access to the external public directory at all? File is selected via SAF and stored inside cgeos _private_ storage
- We should re-request the permission if it should really be necessary

**Comment by andrixnet** [comment]
Apparently this is wider then initially thought.

Anyway, IMO the main is not that c:geo doesn't show a warning/error message that it does not have location access. 
The main problems is that c:geo does not outright request location access as soon as it starts. (as I said, like many other apps when they lose permissions like that). 

Assuming the scenario in which the user actively pressed "deny" location permission, then when the app finishes the startup sequence, show inside the app the very obvious "error - location access denied", preferably as a modal dialog, absolutely not as a bubble message that dissapears silently after a few seconds.

**Comment by fm-sys** [comment]
IMHO it's OK to not show a interrupting warning if the user has actively denied the permission. However if permission is missing without explicit user interaction, the installation wizard should start automatically and request the missing permissions.

**Comment by SammysHP** [comment]
> The main problems is that c:geo does not outright request location access as soon as it starts. (as I said, like many other apps when they lose permissions like that).

This is by design because some people use c:geo without location permission.

**Comment by fm-sys** [comment]
> > The main problems is that c:geo does not outright request location access as soon as it starts. (as I said, like many other apps when they lose permissions like that).
> 
> This is by design because some people use c:geo without location permission.

True. That's why I suggested to differentiate between whether it was users or systems decision.

**Comment by eddiemuc** [comment]
> > > The main problems is that c:geo does not outright request location access as soon as it starts. (as I said, like many other apps when they lose permissions like that).
> > 
> > 
> > This is by design because some people use c:geo without location permission.
> 
> True. That's why I suggested to differentiate between whether it was users or systems decision.

I don't think this can be distinguished. Only thing I can think about is when c:geo stores the current permission setting and checks on startup whether it has changed. 

BTW: there are THREE diffeent location permission settings these days: no permission, coarse location permission and fine location permission

Storage permission seems to be a different problem. Should we introduce EXTERNAL_STORAGE_PERMISSION again for all SDKs?

**Comment by MagpieFourtyTwo** [comment]
> > The main problems is that c:geo does not outright request location access as soon as it starts. (as I said, like many other apps when they lose permissions like that).
> 
> This is by design because some people use c:geo without location permission.

Nevertheless, still possible that there are users which initially denied location access (even possible that this happened by accident and unintentionally), but later on they wonder why it does not work right.

Thus I would *always* at least display the info message in the home screen, no matter if the permission has been actively denied or if it "got lost" by a overprotective system. And additionally there could be a button to trigger c:geo to ask for the permission again. Additionally, same may apply for the storage permission.

**Comment by eddiemuc** [comment]
> Thus I would always at least display the info message in the home screen, no matter if the permission has been actively denied or if it "got lost" by an overprotective system. And additionally there could be a button to trigger c:geo to ask for the permission again. Additionally, same may apply for the storage permission.

Both is available already. Plus, c:geo does not only distinguish between "location permission available" and "unavailable" but also between "coarse-grained" and "fine-grained" location permission. This info is displayed and can be changed via the location bar at the bottom of the home screen currently.

This is how it looks like if you have **only granted coarse-grained**-permission:
![image](https://user-images.githubusercontent.com/6909759/225224207-1e0a7a02-cca5-479c-b7b8-2fd841b8709b.png)

This is how it looks like if you have **no** permission:
![image](https://user-images.githubusercontent.com/6909759/225224434-794b28d5-77f2-4af8-b9f4-36e4928a2e22.png)

When you tap on it you get to the permission-request-dialog:
![image](https://user-images.githubusercontent.com/6909759/225224513-9e38a117-2c35-49f8-9b71-55db5c433f0a.png)
![image](https://user-images.githubusercontent.com/6909759/225224561-7d890d2a-5ef0-4967-bf8e-ac0fe541a786.png)

So the question IMHO is not whether to provide these functions or not, but whether to change their UX design. E.g.:
* Make the warning messages more present (e.g. enlarge them), but in a way which is not too intrinsive for users wishing to use c:geo without fine-grained or any location permission
* Introduce a startup warning if c:geo was stopped with another permission setting than present on startup. E.g. stopped with fine-grained permission, now has coarse-grained permission. Or stopped with no permission, now has coarse-grained permission. Or stopped with fine-grained permission, now has no permission. Exact valid combinations have to be determined.
* ... other ideas

**Comment by MagpieFourtyTwo** [comment]
[Messages for missing location and storage permission]
> Both is available already. 

Perfectly fine! BTW: Would it be possible to add a message for simply switched off GPS, too?

WRT UX design change: I wouldn't put too much effort into this IMHO fairly rare scenario, which moreover typically only happens to "unfaithful" c:geo users anyway. ;) 

No, joking aside: IMO it's quite sufficient to check the current status on startup and output the messages like they seem to be already. Anything more would most probably get a complexity that does not do justice to such a - as said - rather rare problem. And users who consciously decide(d) not to use GPS or not to trust c:geo, should also be aware of what they are doing and that they have to live with messages like these.

**Comment by Lineflyer** [comment]
> No, joking aside: IMO it's quite sufficient to check the current status on startup and output the messages like they seem to be already. Anything more would most probably get a complexity that does not do justice to such a - as said - rather rare problem. And users who consciously decide(d) not to use GPS or not to trust c:geo, should also be aware of what they are doing and that they have to live with messages like these.

Fully agree

**Comment by andrixnet** [comment]
So is this a design decision between 

- expected (akin to most other apps) behaviour towards missing permissions at app startup, that is ask for unavailable permission, then showing warning/error if user actively denied it, thus prioritising active users (new install, wrong permission choice or such system removal) but thus nagging users that intentionally deny some permission.
- accept missing permissions at startup (opposite to expected behaviour), but warning in-app. In this case if using a bubble warning it has very low visibility and fleeting presence. Prioritises users that intentionally deny some permission.

**Comment by MagpieFourtyTwo** [comment]
As said I would just check permissions at startup and display a *static*(!) message in the footer of the home screen. Nothing else.
In fact the way, it is already implemented (if I did not misunderstood @eddiemuc).

**Comment by eddiemuc** [comment]
> As said I would just check permissions at startup and display a *static*(!) message in the footer of the home screen. Nothing else.
> In fact the way, it is already implemented (if I did not misunderstood @eddiemuc).

Correct, that's the way it is currently implemented. Plus permission is re-checked ever time one re-enters the home screen

**Comment by andrixnet** [comment]
I feel this can be closed in the current state.
